src/main/scala/cognite/spark/v1/SdkV1Relation.scala

```diff
@@ -104,37 +106,72 @@ abstract class SdkV1Relation[A <: Product, I](config: RelationConfig, shortName:
       "Update requires an id or externalId to be set for each row.")
     val (updatesById, updatesByExternalId) = updates.partition(_.id.isDefined)
     val updateIds = if (updatesById.isEmpty) { IO.unit } else {
-      resource.updateById(updatesById.map(u => u.id.get -> u.transformInto[U]).toMap)
+      resource
+        .updateById(updatesById.map(u => u.id.get -> u.transformInto[U]).toMap)
+        .flatTap(_ => incMetrics(itemsUpdated, updatesById.size))
     }
     val updateExternalIds = if (updatesByExternalId.isEmpty) { IO.unit } else {
-      resource.updateByExternalId(
-        updatesByExternalId
-          .map(u => u.externalId.get -> u.into[U].withFieldComputed(_.externalId, _ => None).transform)
-          .toMap)
+      resource
+        .updateByExternalId(
+          updatesByExternalId
+            .map(u => u.externalId.get -> u.into[U].withFieldComputed(_.externalId, _ => None).transform)
+            .toMap)
+        .flatTap(_ => incMetrics(itemsUpdated, updatesByExternalId.size))
     }
     (updateIds, updateExternalIds).parMapN((_, _) => ())
   }
 
+  private def assertNoLegacyNameConflicts(e: CdpApiException) = {
+    val legacyNameConflicts =
+      e.duplicated.get.flatMap(j => j("legacyName")).map(_.asString.get)
+    if (legacyNameConflicts.nonEmpty) {
+      throw new IllegalArgumentException(
+        "Found legacyName conflicts, upserts are not supported with legacyName." +
+          s" Conflicting legacyNames: ${legacyNameConflicts.mkString(", ")}")
+    }
+  }
+
   // scalastyle:off no.whitespace.after.left.bracket
-  def upsertAfterConflict[
+  def createOrUpdateByExternalId[
       R <: WithId[Long],
       U <: WithSetExternalId,
       C <: WithExternalId,
       T <: UpdateByExternalId[R, U, IO] with Create[R, C, IO]](
       existingExternalIds: Seq[String],
       resourceCreates: Seq[C],
-      resource: T)(implicit transform: Transformer[C, U]): IO[Unit] = {
+      resource: T,
+      doUpsert: Boolean)(implicit transform: Transformer[C, U]): IO[Unit] = {
     val (resourcesToUpdate, resourcesToCreate) = resourceCreates.partition(
-      p => if (p.externalId.isEmpty) { false } else { existingExternalIds.contains(p.externalId.get) }
+      p => p.externalId.nonEmpty && existingExternalIds.contains(p.externalId.get)
     )
-    val create = if (resourcesToCreate.isEmpty) { IO.unit } else {
-      resource.create(resourcesToCreate)
+    val create = if (resourcesToCreate.isEmpty) {
+      IO.unit
+    } else {
+      resource
+        .create(resourcesToCreate)
+        .flatTap(_ => incMetrics(itemsCreated, resourcesToCreate.size))
+        .handleErrorWith {
+          case e: CdpApiException if doUpsert && e.code == 409 =>
+            assertNoLegacyNameConflicts(e)
+            val existingExternalIds =
+              e.duplicated.get.flatMap(j => j("externalId")).map(_.asString.get)
```

Comment:
Are we sure that `duplicated` is always set here? According to the docs, it can be null, so we might want to map its absence to an empty list instead

Reply:
Fixed by matching on it. Thanks!